text/trace/0247-peer-service-propagation.md

```diff
@@ -0,0 +1,176 @@
+become unreliable, making it eventually obsolete.
+
+This information can be effectively derived in the backend using the
+`Resource` of the parent `Span`, but is otherwise not available
+at Collector processing time, where it could be used for transformation
+purposes or sampling (e.g. adaptive sampling based on the calling service).
+
+As metrics and logs do not have defined a parent-child relationship, using
+`peer.service` could help gaining insight into the remote service as well.
+
+Defining (optional) automated population of `peer.service` will greatly help
+adoption of this attribute by users and vendors explicitly interested in this
+scenario.
+
+## Explanation
+
+SDKs will define an optional feature, disabled by default,
+to read the `service.name` attribute of the related `Resource` and set it
+in the spans' `TraceState` as described in
+[trace state handling](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/tracestate-handling.md)
+specifically using the `us` subkey (denoting **upstream service**):
+
+```
+ot=us:MyService
+```
+
+Instrumentation and processors are then free to use this information to set
+`peer.service` and perform other related processing.
+
+## Internal details
+
+SDKs will disable by default this option, maintaining the current behavior.
+When the feature is explicitly enabled by the user, spans will include
+an additional entry in `TraceState` as described above. By doing this,
+the user acknowledges the additional cost in memory and bandwidth.
+
+Span creation will be updated like this:
+
+```java
+//
+// SpanBuilder.startSpan()
+//
+if (tracerSharedState.propagateServiceName) {
+  String serviceName = tracerSharedState.getResource().getAttribute(PEER_SERVICE);
+  traceState = addServiceNameToTracerState(traceState, serviceName);
+}
+// Use the updated `traceState` to create the new SpanContext.
+```
+
+Server-side instrumentation (e.g. http servers, gRPC on the receiver side)
+can then be updated to use the propagated context to look for the `us` subkey
+in `TraceState`, and if it exists, use it to set `peer.service` on the local `Span`:
+
+```java
+// 
+// Incoming request handling.
+//
+try (Scope scope = extractRemoteContext(headers).makeCurrent()) {
+  SpanBuilder spanBuilder = tracer.spanBuilder("server-span");
+  
+  TraceState remoteTraceState = Span.current()
+      .getSpanContext()
+      .getTraceState();
+  String peerServiceName = getUpstreamServiceName(remoteTraceState);
+  if (peerServiceName != null) {
+    spanBuilder.setAttribute(PEER_SERVICE, peerServiceName);
+  }
+}
+```
+
+With `peer.service` present in server spans, further processing, filtering and sampling can
+then be accomplished in the Collector, e.g. a preview of the dependency map of a service,
+similar in spirit to zPages could be created.
+
+### Use scenarios
+
+Sampling can benefit from knowing the calling service, specifically:
```

Comment:
As sampling is mentioned as a primary usage scenario for this feature, I wonder if would make sense to bundle this together with other sampling related values for which there's a proposal to propagate them via trace state: https://github.com/open-telemetry/oteps/pull/235

All those values could then be consistently used and populated by samplers, one wouldn't need to invent a new configurable mechanism in SDKs (at least for the client side).

Reply:
This value makes sense to propagate with the sampling-related attributes in the `tracestate`, which are now covered in https://github.com/open-telemetry/semantic-conventions/pull/793. Still, I see it as an independent property.